### PR TITLE
Support "format" field on "stackTrace" request

### PR DESCRIPTION
**Pull request by andrewcrawley**
Adds support for the "value formatting" feature of DAP, which allows flags to be passed to control formatting of stack frames, as well as variables and evaluation results.

Note that while this commit wires up the DAP formatting flags to the relevant AD7 calls, MIEngine currently doesn't make use of the radix value passed, so the flag to control hex vs decimal formatting doesn't work.

### CODE REVIEW

File: src/OpenDebugAD7/AD7DebugSession.cs [hunk 1]
```diff
@@ -1274,31 +1272,77 @@ protected override void HandleStackTraceRequestAsync(IRequestResponder<StackTrac
...
-                            frameEnumInfo = new ThreadFrameEnumInfo(frameEnum, totalFrames);
-                            m_threadFrameEnumInfos.Add(threadReference, frameEnumInfo);
+                            // No formatting flags provided in the request - use the default format, which includes the module name and argument names / types
+                            flags |= enum_FRAMEINFO_FLAGS.FIF_FUNCNAME_MODULE |
+                                     enum_FRAMEINFO_FLAGS.FIF_FUNCNAME_ARGS |
+                                     enum_FRAMEINFO_FLAGS.FIF_FUNCNAME_ARGS_TYPES |
+                                     enum_FRAMEINFO_FLAGS.FIF_FUNCNAME_ARGS_NAMES;
                         }
+
+                        thread.EnumFrameInfo(flags, radix, out IEnumDebugFrameInfo2 frameEnum);
```
Comment (gregg-miskelly) [inline src/OpenDebugAD7/AD7DebugSession.cs:1336]: >thread [](start = 24, length = 6)

Looks like we had this problem before also, but: check hr?

**Comment by gregg-miskelly** [comment]
>             property.GetPropertyInfo(propertyInfoFlags, Constants.EvaluationRadix, Constants.EvaluationTimeout, null, 0, propertyInfo);

Should we wire up this one too, or at least leave a TODO comment?

---
Refers to: src/OpenDebugAD7/AD7DebugSession.cs:2270 in 7375184. [](commit_id = 73751849ddfec619d50f5af2cda445f503c22800, deletion_comment = False)

**Comment by gregg-miskelly** [comment]
Sorry if you already know this, but if we want this to really work, we will need to have OpenDebugAD7 start calling SetMainThreadSettingsCallback110, and implement IDebugSettingsCallback110.GetDisplayRadix.

**Comment by andrewcrawley** [comment]
> 
> 
> Sorry if you already know this, but if we want the radix stuff to work, we will need to have OpenDebugAD7 start calling SetMainThreadSettingsCallback110, and implement IDebugSettingsCallback110.GetDisplayRadix.

I considered this, but I was concerned about race conditions when (temporarily?) changing global state in response to an async DAP request.  It's common to simultaneously issue several `stackTrace` requests with different formatting options, for instance.  Is there a reason MIEngine always refers to the global setting instead of using the radix passed via the individual APIs?

**Comment by gregg-miskelly** [comment]
> Is there a reason MIEngine always refers to the global setting instead of using the radix passed via the individual APIs?
I believe it was done this way because:
- This is a global setting in GDB, so the race would be there no matter what
- The poor design of the original AD7 evaluation APIs where display radix isn't passed until evaluate sync
- In VS, there isn't much of a reason not to do it that way since there really is a global setting

I suspect we could get away with using the global setting if we just were careful to only update the global setting in cases where the request actually needs it. For example, a request to obtain stack frames that doesn't want values doesn't need to know the radix.

File: src/OpenDebugAD7/AD7DebugSession.cs [hunk 2]
```diff
@@ -1274,31 +1282,85 @@ protected override void HandleStackTraceRequestAsync(IRequestResponder<StackTrac
             // Make sure we are stopped and receiving valid input or else return an empty stack trace
```
Comment (gregg-miskelly) [inline src/OpenDebugAD7/AD7DebugSession.cs:1282]: I think this method needs a try/catch to turn exceptions into error responses.

File: src/OpenDebugAD7/AD7DebugSession.cs [hunk 3]
```diff
@@ -2211,7 +2291,25 @@ protected override void HandleEvaluateRequestAsync(IRequestResponder<EvaluateArg
...
+                if (format.Hex == true)
+                {
+                    radix = 16;
+                }
+            }
+
+            if (m_settingsCallback != null)
+            {
+                // MIEngine generally gets the radix from IDebugSettingsCallback110 rather than using the radix passed
+                m_settingsCallback.Radix = radix;
```
Comment (gregg-miskelly) [inline src/OpenDebugAD7/AD7DebugSession.cs:2309]: This needs to be done earlier. I would probably do it before making any calls into the engine, but definitely before EvaluateSync.

**Review by gregg-miskelly — APPROVED** [review]
Otherwise LGTM

**Review by gregg-miskelly — APPROVED** [review]

**Pull request merged**